# Импорты

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [28]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import shutil


%matplotlib inline

In [3]:
pd.options.display.max_columns = 50
pd.options.display.max_rows = 50
pd.options.display.max_colwidth = 200
pd.options.display.width = 200

In [4]:
path_to_folder = '/content/drive/MyDrive/internship_fintech/dataset'

In [5]:
CLASSES = {1: 'Надежный клиент',
           0: 'Ненадежный клиент'}

# Предобработка данных

In [29]:
data = pd.read_excel(f'{path_to_folder}/dataset.xlsx')
data.head()

(167105, 41)


region_name      city_name  population date of birth   gender education                  position    dt_nbki                               id_nbki     issued    amount currency  \
0  Новосибирская область  г.  Барабинск     29122.0    1980-07-28  Женский    Высшее  работник бюджетной сферы 2023-03-05  27ab5815-bb2d-11ed-b80d-3cfdfed12971 2013-10-07   75000.0      RUB   
1  Новосибирская область  г.  Барабинск     29122.0    1980-07-28  Женский    Высшее  работник бюджетной сферы 2023-03-05  27ab5815-bb2d-11ed-b80d-3cfdfed12971 2013-11-06  155740.0      RUB   
2  Новосибирская область  г.  Барабинск     29122.0    1980-07-28  Женский    Высшее  работник бюджетной сферы 2023-03-05  27ab5815-bb2d-11ed-b80d-3cfdfed12971 2016-04-05    5000.0      RUB   
3  Новосибирская область  г.  Барабинск     29122.0    1980-07-28  Женский    Высшее  работник бюджетной сферы 2023-03-05  27ab5815-bb2d-11ed-b80d-3cfdfed12971 2016-04-16    5000.0      RUB   
4  Новосибирская область  г.  Барабинск     29122.0    1980-07-28  Женский    Высшее  работник бюджетной сферы 2023-03-05  27ab5815-bb2d-11ed-b80d-3cfdfed12971 2016-05-13    5000.0      RUB   

       type_credit status_credit               period period_interest     psk   balance  current_indebtedness  expired days_overdue  next_payment last_payment    updated  \
0   КредитнаяКарта     Просрочен  3999-12-31 00:00:00             NaT  34.927  149997.8                   0.0      0.0          NaN           0.0   2019-11-29 2023-01-12   
1  Потребительский     Просрочен  2014-12-31 00:00:00             NaT  41.715   66261.2                   0.0      0.0          NaN           0.0   2019-11-29 2023-01-12   
2      Микрокредит    СчетЗакрыт  2016-05-05 00:00:00             NaT   0.000    6000.0                   0.0      0.0          NaN           0.0   2016-04-14 2016-04-14   
3      Микрокредит    СчетЗакрыт  2016-05-27 00:00:00             NaT   0.000    7700.0                   0.0      0.0          NaN           0.0   2016-05-12 2016-05-12   
4      Микрокредит    СчетЗакрыт  2016-06-25 00:00:00             NaT   0.000    8000.0                   0.0      0.0          NaN           0.0   2016-06-11 2016-06-11   

   term_indebtedness_body_loan  term_indebtedness_interest  term_indebtedness_other  overdue_indebtedness_body_loan  overdue_indebtedness_interest  overdue_indebtedness_other  \
0                          0.0                         0.0                  4965.70                        73721.52                        7510.60                     7227.78   
1                          0.0                         0.0                 17885.76                       145740.27                        4211.84                   114528.02   
2                          0.0                         0.0                     0.00                            0.00                           0.00                        0.00   
3                          0.0                         0.0                     0.00                            0.00                           0.00                        0.00   
4                          0.0                         0.0                     0.00                            0.00                           0.00                        0.00   

   actual_payments_24_Months volume_commitment  amount_surety surety_period source_information  maximum_amount_overdue_indebtedness date_credit_state beginning_days_delinquency_counting type_debt  \
0                        0.0               NaN            0.0           NaT               Банк                                  0.0        2023-01-12                          2014-09-13       NaN   
1                        0.0               NaN            0.0           NaT               Банк                                  0.0        2023-01-12                          2014-09-02       NaN   
2                        0.0               NaN            0.0           NaT                МФО                                  0.0        2016-04-14

In [30]:
# Добавим единичный столбец target (надежные клиенты)
target_col = [1 for i in range(data.shape[0])]
data.insert(loc=len(data.columns) , column='target', value=target_col)

In [31]:
print('Размеры датасета', data.shape)
print('Общее количество пропусков:', data.isna().sum().sum(), 'или', round(data.isna().sum().sum() / (data.shape[0]*data.shape[1])*100, 2), '%\n')
print(f'Доли пропусков по столбцам:\n{round(data.isna().sum() / data.shape[0] * 100, 4)}')
print(data.info())
round(data.describe(), 1)

Размеры датасета (167105, 42)
Общее количество пропусков: 1031722 или 14.7 %

Доли пропусков по столбцам:
region_name                             0.0132
city_name                               0.0132
population                              0.0132
date of birth                           0.0000
gender                                  0.0000
education                              28.1039
position                                8.8208
dt_nbki                                 0.0000
id_nbki                                 0.0000
issued                                  0.0269
amount                                  0.0269
currency                                0.0000
type_credit                             0.0389
status_credit                           0.0275
period                                  0.0269
period_interest                        95.0133
psk                                     0.0269
balance                                 0.0269
current_indebtedness                    0.0269
e

population      amount       psk     balance  current_indebtedness    expired  next_payment  term_indebtedness_body_loan  term_indebtedness_interest  term_indebtedness_other  \
count    167083.0    167060.0  167060.0    167060.0              167060.0   167060.0      167060.0                     167060.0                    167060.0                 167060.0   
mean     211702.9     22393.4     317.2     24996.2                 405.8      188.1         130.4                       5610.1                       149.0                     86.5   
std      466756.1     93555.3    2100.1    106599.2               12070.6     9108.2        6501.8                      55943.6                      2013.7                   4667.2   
min        3237.0         0.0       0.0         0.0                   0.0        0.0           0.0                          0.0                         0.0                      0.0   
25%       32592.0      5000.0     237.2      3780.0                   0.0        0.0           0.0                          0.0                         0.0                      0.0   
50%      126805.0     10000.0     364.9      9260.5                   0.0        0.0           0.0                          0.0                         0.0                      0.0   
75%      182291.0     15919.9     365.0     17984.1                   0.0        0.0           0.0                          0.0                         0.0                      0.0   
max     5383890.0  13090615.8  844793.0  13779070.3             2125788.0  2125788.0     2125788.0                    6298927.0                    182559.1                1258251.7   

       overdue_indebtedness_body_loan  overdue_indebtedness_interest  overdue_indebtedness_other  actual_payments_24_Months  amount_surety  maximum_amount_overdue_indebtedness  \
count                        167060.0                       167060.0                    167060.0                   167060.0       167060.0                             167060.0   
mean                           1292.9                          403.3                        77.9                      288.2          206.6                                514.9   
std                           18677.1                         9562.8                      2597.9                     6909.5        25674.5                              10778.0   
min                               0.0                            0.0                         0.0                        0.0            0.0                                  0.0   
25%                               0.0                            0.0                         0.0                        0.0            0.0                                  0.0   
50%                               0.0                            0.0                         0.0                        0.0            0.0                                  0.0   
75%                               0.0                            0.0                         0.0                        0.0            0.0                                  0.0   
max                         2622570.0                      1430287.3                    500000.0                   979349.0      6418250.0                            1572856.7   

       amount_principal_debt_repaid  availability_support    target  
count                      167060.0              167060.0  167105.0  
mean                        17234.0                   0.0       1.0  
std                         86793.6                   0.1       0.0  
min                             0.0                   0.0       1.0  
25%                          2000.0                   0.0       1.0  
50%                          6719.9                   0.0       1.0  
75%                         13330.1                   0.0       1.0  
max                      13674558.4                   1.0       1.0

**1. Столбец period не датетайм а обджект**
 

**2. Есть пустые данные**
Но столбцы нельзя удалять, т.к. могут прийти данные с пустыми значениями. Или можно удалить столбцы?

- education уровень образования 28.1 % изменить на моду
- period_interest Срок (проценты) 95.01 % NaT Заменить на ?
- days_overdue Количество дней просрочки 94.23 % NaN Заменить на 0
- last_payment Последний платеж 10.6 % NaT Заменить на ?
- volume_commitment Объем обязательства 99.99 % NaN Заменить на 0
- surety_period Срок поручительства 99.98 % NaT Заменить на ?
- beginning_days_delinquency_counting Начало отсчета дней просрочки 90.61 % NaT Заменить на ?
- type_debt Тип задолженности 89.34 % NaN Заменить на ?

**Стандартизация и преобразование в OHE?**

**Нужно ли преобразовать даты?**

## Теперь нужно просмотреть датасет на здравый смысл. Чтобы не было 3999 года, кредита на 1000 руб.

In [ ]:
# def clear_data(data):
#     data = data.drop(['period_interest',
#                   'Количество дней просрочки',
#                   'Объем обязательства',
#                   'Срок поручительства',
#                   'Начало отсчета дней просрочки',
#                   'Тип задолженности'], axis=0)


#     data['образование'] = data['образование'].apply(lambda x: x if x is not np.NaN else 'Unknown')
#     data['должность'] = data['должность'].apply(lambda x: x if x is not np.NaN else 'Unknown')

#     last_payment_median = data['Последний платеж'].median()
#     data['Последний платеж'] = data['Последний платеж'].apply(lambda x: x if x is not pd.NaT else last_payment_median)

#     data = data[data['Валюта']!='USD']

#     data = data.dropna()

    # print('Размеры датасета', data.shape)
    # print('\nОбщее количество пропусков:', data.isna().sum().sum(), 'или', round(data.isna().sum().sum() / (data.shape[0]*data.shape[1])*100, 2), '%\n')
   
    # return data

## Кодирование категориальных данных

**Есть категориальные данные. Необходимо преобразовать в числовые**

In [27]:
cat_columns = data.select_dtypes(['object']).columns
cat_nunique = {i: data[i].nunique() for i in cat_columns}
cat_nunique

{'region_name': 70,
 'city_name': 747,
 'gender': 2,
 'education': 7,
 'position': 19,
 'id_nbki': 5000,
 'currency': 2,
 'type_credit': 17,
 'status_credit': 12,
 'period': 6780,
 'days_overdue': 1708,
 'volume_commitment': 1,
 'source_information': 8,
 'type_debt': 2}

In [ ]:
# data[cat_columns] = data[cat_columns].apply(lambda x: pd.factorize(x)[0])
# print(len(data.select_dtypes(['object']).columns) == 0)
# data.info()

In [ ]:
# for i in data.columns:
#     print(f'column: {i}')
#     print(f'min {np.min(data[i])}')
#     print(f'max: {np.max(data[i])}')
#     print(f'99%: {np.percentile(data[i], 99)}')
#     print(f'97%: {np.percentile(data[i], 97)}')
#     print(f'90%: {np.percentile(data[i], 90)}')
#     print(f'median: {np.percentile(data[i], 50)}')
#     print('-------------'*2)

## Корреляции

* Корреляция - коэффициент, показывающий степень зависимости двух факторов (-1,1)

* Необходимо чтобы корреляция была высокой между факторами и ЦФ (price), если есть несколько факторов - оставить один

Провести анализ feature selection - выкинуть ненужные, неважные факторы

## Визуализация

In [ ]:
# plt.figure(figsize=(10, 6))
# plt.hist(data., bins=50, ec='black', color='#2196f3')
# plt.xlabel('Цена')
# plt.ylabel('Кол-во домов')
# plt.show()